extensions/interactions/MathExpressionInput/MathExpressionInput.js

```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+
+          /**
+           * Adds a button overlay and invisible text field used to bring up
+           * the keyboard on mobile devices.
+           *
+           * TODO: On iOS/Safari, keyboard may only appear on first press.
+           * This may not be a significant issue as the MathExpressionInput is
+           * recreated if the given answer is incorrect.
+           */
+          var addMobileButtonOverlay = function() {
+            /**
+             * Checks if the guppy div has a width and height greater than 0,
+             * if not schedules a timeout to run again after 100ms. If the
+             * guppy div has a valid width/height, position the invisible
+             * button directly over the guppy div. For mobile browsers, focus()
+             * can only be called from within an onclick handler. Using a form
+             * element was more reliable than attaching the handler to the
+             * guppy div directly.
+             */
+            var positionButtonOverlay = function() {
+              var guppyOffset = $(guppyDivElt).position();
+              var guppySize = document.querySelector('.guppy-div')
+                .getBoundingClientRect();
+              console.log(guppySize, $('#startMathInputButton'));
+
+              // If the guppy div hasn't rendered yet, retry after 100ms.
+              if (guppySize.width === 0 || guppySize.height === 0) {
+                $timeout(positionButtonOverlay, 100);
+              } else {
+                $('#startMathInputButton').css({
+                  top: guppyOffset.top,
+                  left: guppyOffset.left,
+                  width: guppySize.width,
+                  height: guppySize.height
+                });
+              }
+            };
+            positionButtonOverlay();
+
+            // The focus() call must be in a click event handler and on a text
+            // field to make the mobile keyboard appear.
+            $('#startMathInputButton').click(function() {
+              guppyInstance.activate();
+
+              var fakeInputElement = document.
+                querySelector('#fakeInputForMathExpression');
+              fakeInputElement.focus();
+
+              // Place the cursor at the end of the text input, so that the
+              // user can use backspace to delete.
+              fakeInputElement.setSelectionRange(fakeInputElement.value.length,
+                fakeInputElement.value.length);
+            });
+
+            // Mapping characters back to Mousetrap codes.
+            // Use Mousetrap.trigger(code) to replay them.
+            var K_SYM_REVERSE_MAP = {
+              '^': 'shift+6',
+              '*': 'shift+8',
+              '(': 'shift+9',
+              '<': 'shift+,',
+              '>': 'shift+.',
+              '\\': 'shift+\\',
+              ' ': 'space',
+              ')': 'shift+0'
+              // Missing up/down. user needs to use [space], 'sub' instead.
```

Comment:
In the standard Guppy, the user could for example press the down arrow to go from entering an exponent to a base level variable/number, or from base to a subscript. Since those are not printable characters, this method does not proxy (replay) them to Guppy.

On a desktop browser, if you type "frac" you can then use the arrow keys to position the cursor in the top or bottom box of the fraction. On mobile, you would need to type "fraca b" to get a fraction with a as the numerator and b the denominator. *This works the same as it would on desktop, it's only that desktop users are more likely to use arrow keys than spaces.

Mobile devices typically do not have arrow key input anyways though - but still this is an area where it is likely to be unintuitive for users. How would a typical user find out that they need to press [space] to get out of things like exponents and fractions, and write out 'sub' to start a subscript? (Among other functions listed [here.](https://daniel3735928559.github.io/guppy/))